vms-admitter: remove validation of vm clone volume from webhook

ShellyKa13 opened the pull request:
When having this check in the webhook we are preventing the creation of VMs which their datasource might not exist yet. Furthermore there are cases like in restore of backup where the vms volumes are already populated and dont event need to create the datavolume hence for sure no need to prevent the VM from being created in absence of such source.
This fix aims to fix bug: https://issues.redhat.com/browse/CNV-43272 While also make our product more aligned with eventual consistency model.

**Note: to clarify the removed authentication from the webhook is still being done in the vm reconcile loop before creating the datavolumes**

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes # Jira-ticker: https://issues.redhat.com/browse/CNV-43272

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

ShellyKa13 commented:
/sig storage
/sig code-quality

ShellyKa13 commented:
/cc @mhenriks

mhenriks commented:
/hold

Let's get #12547 in first so we have a PR that we can easily backport 

Then this one will only hit main and release 1.4+

ShellyKa13 commented:
/unhold
@mhenriks I wrote to you also in the PR you mentioned #12547 
I think this PR fixes several issues related to creating a vm with dvtemplate with clone source which the other PR doesnt solve, hence I think we can and should backport this PR instead and drop PR #12547

mhenriks commented:
@ShellyKa13 Not good form for you to unhold a PR you did not put a hold on

ShellyKa13 commented:
/hold
You are right. Sorry.

ShellyKa13 commented:
@mhenriks as agreed PR #12547 was merged and will be backported to previous releases.
I added as a first commit a revert of it since it added logic to the auth check which I removed in this PR.
Can you please review this PR again and see if it looks good to you?

ShellyKa13 commented:
/retest

ShellyKa13 commented:
/retest

mhenriks commented:
Looks okay but clearly some functests need to be updated

Reverting #12547 in it's entirety could be considered a little heavy handed, but okay I guess.

Debating whether we want to wait for 1.5, kind of big change before a release. What do you think?

ShellyKa13 commented:
@mhenriks I looked at the revert commit and I did see some code which has value and not directly related to the auth check so I edited the revert commit and removed only the code related to the auth check, thanks for pointing that out!
I fixed the functests.
Regarding the release, I would prefer to have this change in the upcoming release, I believe it doesnt present a risk to the user or a risk in the stability of the release, also I believe it has a value in fixing a bug which I believe should be fixed as fast as possible and not be pushed for another period of time. Of course if you think otherwise I respect your opinion and am willing to wait with this, so up to you :)

ShellyKa13 commented:
@mhenriks after checking the original bug that I opened the PR for I believe your fix https://github.com/kubevirt/kubevirt/pull/12547 already solves that issue. So this PR is only a nice to have in sake of consistency. So if you think we should wait with it I dont mind.

mhenriks commented:
> @mhenriks after checking the original bug that I opened the PR for I believe your fix #12547 already solves that issue. So this PR is only a nice to have in sake of consistency. So if you think we should wait with it I dont mind.

I think I may not mind getting this in for 1.4. Any thought @alicefr?

mhenriks commented on tests/storage/datavolume.go:
```diff
@@ -985,9 +986,14 @@ var _ = SIGDescribe("DataVolume Integration", func() {
+				vm, err := virtClient.VirtualMachine(vm.Namespace).Create(context.Background(), vm, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				defer func() {
+					err := virtClient.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{})
```
Can you remind me why this defer is necessary?

ShellyKa13 replied:
not sure why I added it, can't think of a reason why would it be needed since the vm should be cleaned up when the test ends anyways.. Will remove it..

ShellyKa13 replied:
done

mhenriks commented:
/retest-required

mhenriks commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

mhenriks commented:
/approve

EdDev commented:
@arnongilboa , can you please have a look at this change?

arnongilboa commented:
> @arnongilboa , can you please have a look at this change?

sure, it's in my queue

arnongilboa commented on tests/storage/datavolume.go:
```diff
@@ -1025,30 +1025,6 @@ var _ = SIGDescribe("DataVolume Integration", func() {
 				Entry("with explicit role (one namespace) snapshot clone", explicitCloneRole, false, true, snapshotCloneMutateFunc, false),
 			)
-
-			It("should skip authorization when DataVolume already exists", func() {
```
Why is this test irrelevant anymore? shouldn't we test it even after dropping the webhook validation?

ShellyKa13 replied:
This was part of a PR of Micheal which is mostly since he added this logic to the webhook, but I guess I can keep this test since it can test the vm watch as well

arnongilboa commented on pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:
```diff
@@ -1602,48 +1594,6 @@ var _ = Describe("Validating VM Admitter", func() {
 			Entry("when everything suppied with 'sa' service account", "ns1", "ns2", "ns3", "sa", "ns3", "ns2", "sa"),
 		)
 
-		It("should successfully authorize clone with existing DataVolume", func() {
```
Why is this utest irrelevant anymore? shouldn't we test it even after dropping the webhook validation?

ShellyKa13 replied:
the auth is checked in the vm watch ut. no need to test it here anymore since the webhook doesnt take it into account anymore

arnongilboa commented on pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:
```diff
@@ -1537,195 +1532,6 @@ var _ = Describe("Validating VM Admitter", func() {
 			Expect(causes[0].Field).To(Equal("fake"))
 		})
-
-		DescribeTable("should successfully authorize clone", func(arNamespace, vmNamespace, sourceNamespace,
```
Why is this utest irrelevant anymore? shouldn't we check clone authorization from same / differnt ns?

ShellyKa13 replied:
This is being tested in the vm watch

arnongilboa commented on tests/storage/datavolume.go:
```diff
@@ -1002,18 +1002,11 @@ var _ = SIGDescribe("DataVolume Integration", func() {
-				// stop vm
-				vm = libvmops.StopVirtualMachine(vm)
+				libvmops.StartVirtualMachine(vm)
 			},
```
intentionally removed the DV succeeded check covered by createVMSuccess?no need to stop as the ns is deleted?

ShellyKa13 replied:
I dont need to create the vm anymore since the webhook doesnt fail the creation, instead the VM is created I see that the condition I expect is there and then I start the vm in that function there is a check the VM started successfully which assures the dv succeeded.

arnongilboa commented:
/lgtm

ShellyKa13 commented:
/retest
